liblwgeom/lwgeom_geos.c

```diff
@@ -1233,6 +1244,83 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
 	return result;
 }
 
+static LWGEOM*
+lwgeom_offsetcurve_multilinestring(const LWGEOM* geom, double size, int quadsegs, int joinStyle, double mitreLimit)
+{
+	int32_t srid = get_result_srid(geom, NULL, __func__);
+	uint8_t is3d = FLAGS_GET_Z(geom->flags);
+	if (srid == SRID_INVALID) return NULL;
+
+	if (!lwgeom_is_collection(geom))
+		return lwgeom_offsetcurve_line(lwgeom_as_lwline(geom), size, quadsegs, joinStyle, mitreLimit);
+	else
+	{
+		LWCOLLECTION* result;
+		LWGEOM* tmp;
+		const LWCOLLECTION* col = lwgeom_as_lwcollection(geom);
+		int i;
+		result = lwcollection_construct_empty(MULTILINETYPE, srid, is3d, LW_FALSE);
+		for (i = 0; i < col->ngeoms; i++)
+		{
+			tmp = lwgeom_offsetcurve_line(
+			    lwgeom_as_lwline(col->geoms[i]), size, quadsegs, joinStyle, mitreLimit);
+
+			if (!tmp)
+			{
+				lwcollection_free(result);
+				return NULL;
+			}
+
+			if (lwgeom_get_type(tmp) != LINETYPE && lwgeom_get_type(tmp) != MULTILINETYPE) return NULL;
+
+			if (!lwgeom_is_empty(tmp))
+			{
+				if (!lwgeom_is_collection(tmp))
+					lwcollection_add_lwgeom(result, tmp);
+				else
+				{
+					const LWCOLLECTION* c2 = lwgeom_as_lwcollection(tmp);
+					int j;
+					for (j = 0; j < c2->ngeoms; j++)
+						lwcollection_add_lwgeom(result, c2->geoms[j]);
+					lwgeom_release(tmp);
+				}
+			}
+		}
+		if (result->ngeoms == 1)
+		{
+			tmp = result->geoms[0];
+			lwcollection_release(result);
+			return tmp;
+		}
+		else
+			return lwcollection_as_lwgeom(result);
+	}
+}
+
+LWGEOM*
+lwgeom_offsetcurve(const LWGEOM* geom, double size, int quadsegs, int joinStyle, double mitreLimit)
+{
+	int32_t srid = get_result_srid(geom, NULL, __func__);
+	LWGEOM* result, *tmp;
+	if (srid == SRID_INVALID) return NULL;
+
+	result = lwgeom_offsetcurve_multilinestring(geom, size, quadsegs, joinStyle, mitreLimit);
+	if (!result)
+	{
+		LWGEOM* noded = lwgeom_node(geom);
+		if (noded)
+		{
+			result = lwgeom_offsetcurve_multilinestring(noded, size, quadsegs, joinStyle, mitreLimit);
+			lwfree(noded);
+			if (!result) lwerror("lwgeom_offsetcurve: noding input still does not let it be processed.");
+		}
+	}
+	tmp = lwgeom_make_valid(result);
```

Comment:
May need to adjust implicit makevalid based on closure of dev list discussion

Reply:
removed makevalid for now